backend/src/openarchiefbeheer/zaken/utils.py

```diff
@@ -136,3 +145,141 @@ def retrieve_selectielijstklasse_choices(process_type_url: str) -> list:
+            execute_unless_result_exist(
+                result_store,
+                "besluiten",
+                decision["url"],
+                partial(brc_client.delete, f"besluiten/{decision_uuid}"),
+            )
+
+
+def handle_document_delete_exc(exc: HTTPError) -> None:
+    response = exc.response
+    if not response.status_code == status.HTTP_204_NO_CONTENT and not (
+        response.status_code == status.HTTP_400_BAD_REQUEST
+        and response.json()["invalidParams"][0]["code"] == "pending-relations"
+    ):
+        response.raise_for_status()
+
+
+def delete_documents(
+    zaak: "Zaak", zrc_client: APIClient, result_store: ResultStore
+) -> None:
+    with zrc_client:
+        response = zrc_client.get("zaakinformatieobjecten", params={"zaak": zaak.url})
+        response.raise_for_status()
+        zios = response.json()
+        if not len(zios) and not result_store.has_resource_to_delete(
+            "enkelvoudiginformatieobjecten"
+        ):
+            return
+
+        for zio in zios:
+            result_store.add_resource_to_delete(
+                "enkelvoudiginformatieobjecten", zio["informatieobject"]
+            )
+            zio_uuid = furl(zio["url"]).path.segments[-1]
+            execute_unless_result_exist(
+                result_store,
+                "zaakinformatieobjecten",
+                zio["url"],
+                partial(zrc_client.delete, f"zaakinformatieobjecten/{zio_uuid}"),
+            )
+
+    drc_service = Service.objects.get(api_type=APITypes.drc)
+    drc_client = build_client(drc_service)
+
+    with drc_client:
+        for document_url in result_store.get_resources_to_delete(
+            "enkelvoudiginformatieobjecten"
+        ):
+            document_uuid = furl(document_url).path.segments[-1]
+            execute_unless_result_exist(
+                result_store,
+                "enkelvoudiginformatieobjecten",
+                document_url,
+                partial(
+                    drc_client.delete, f"enkelvoudiginformatieobjecten/{document_uuid}"
+                ),
+                handle_document_delete_exc,
+            )
+
+    result_store.clear_resources_to_delete("enkelvoudiginformatieobjecten")
+
+
+def delete_zaak(zaak: "Zaak", zrc_client: APIClient, result_store: ResultStore) -> None:
+    with zrc_client:
+        execute_unless_result_exist(
+            result_store,
+            "zaken",
+            zaak.url,
+            partial(zrc_client.delete, f"zaken/{zaak.uuid}"),
+        )
+
+
+def delete_zaak_and_related_objects(zaak: "Zaak", result_store: ResultStore) -> None:
+    """Delete a zaak and related objects
+
+    The procedure to delete all objects related to a zaak is as follows:
+    - Check if there are besluiten (in the Besluiten API) related to the zaak and delete them. This automatically deletes
+      ZaakBesluiten in the Zaken API.
+      Q: does it automatically delete BIOs and Documents related to the deleted besluit?
```

Comment:
I haven't been able to test this because of https://github.com/open-zaak/open-zaak/issues/1710